docs: document partition manager predefined configurations (#766)
NordicPlayground/fw-nrfconnect-nrf: doc/nrf/ug_partition_manager.rst

      --slot-size $<TARGET_PROPERTY:partition_manager,MCUBOOT_SLOT_SIZE>

Predefined configuration
========================

> Predefined configuration is a feature intended for deployed products which consists of multiple images, but where

Reviewer: ```suggestion
The predefined configuration is a feature intended for deployed products which consist of multiple images, but where
only a subset of the included images will be upgradeable through firmware update mechanisms.
In these use cases, it is crucial that the images to be upgraded are linked to the same address as what is on the existing product.
This to avoid accidental overrides and incorrect configurations.
Within the upgradeable area, the predefined configuration also allows defining new partitions, deleting existing partitions, and modifying existing partitions.
```

only a subset of the included images will be upgrade-able through firmware update mechanisms.

> In this case, it is important that the images to be upgraded are linked to the same address as what is on the existing product.

Reviewer 2: For clarity:
"In this case" -> "In this use case"

Author: Made it "In these use cases"

This to avoid accidental overrides, and incorrect configurations.

> Predefined configuration also allow defining new partitions, deleting existing partitions, and re-sizing existing partitions.

Reviewer 2: "Predefined ..." -> "Within the upgradeable area, predefined ..."

> Within the upgradeable area, predefined configuration also allow defining new partitions, deleting existing partitions, and modifying existing partitions.

Reviewer: The predefined configuration is a feature intended for deployed products which consist of multiple images, but where
only a subset of the included images will be upgradeable through firmware update mechanisms.
In these use cases, it is crucial that the images to be upgraded are linked to the same address as what is on the existing product.
This to avoid accidental overrides and incorrect configurations.
Within the upgradeable area, the predefined configuration also allows defining new partitions, deleting existing partitions, and modifying existing partitions.

Providing predefined configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The predefined configuration is provided through a YAML formatted configuration file.
This file is similar to the regular :file:`pm.yml` configuration files, except
that it also defines the start address and size of all partitions.
The given configuration for a build can be found in :file:`${BUILD_DIR}/zephyr/include/generated/partitions.yml`.
To apply the current configuration as a predefined configuration, copy this file to the application source directory.
When the build system sees a file named :file:`partitions.yml` in an applications source directory it is automatically provided to the partition manager script as the predefined configuration.

> 

Reviewer 2: Add something like: "If desired, the file can be modified as described below."

> When the build system sees a file named :file:`partitions.yml` in an applications source directory it is automatically provided to the partition manager script as the predefined configuration.

Reviewer: ```suggestion
When the build system sees a file named :file:`partitions.yml` in an applications source directory, it is automatically provided to the partition manager script as the predefined configuration.
```

Removing a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
A predefined partition is removed by deleting its entry in :file:`partitions.yml`.
Note that only partitions adjacent with the ``app`` partition can be deleted.

> Re-sizing a predefined partition

Reviewer 2: How about: "Re-sizing" -> "Modifying" in this paragraph?

> Note that only partitions adjacent with the ``app`` partition can be deleted.

Reviewer 2: Note that only partitions adjacent to the ``app`` partition or other modified/deleted partitions can itself be deleted.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To re-size a predefined partition, first remove the partition from :file:`partitions.yml` then ensure that the build strategy for that sub image is "Build from source" and update the configured partition size.

Reviewer 2: 'the build startegy is "Build from source"' is quite idiomatic. How about something like 'the partition will be re-created from the existing "pm.yml"'?

Author: The reason for mentioning the build strategy is that the user will typically have this set to either "Skip" or "Include Hex File", in which case the user actively has to change this config to "build from source" in order to evaluate the pm.yml

> Note that only partitions adjacent with the ``app`` partition can be re-sized.

Reviewer 2: This "adjacent" rule is not quite correct since (I think) partitions adjacent to other modified partitions can be modified as well. Can you clarify this?

Author: My thought was that in the situation "A B APP C" one will first remove B to get "A APP C" and then A can be modified as well.

It might be useful to define the concept of "upgrade-able area" as you mentioned in one of your earlier comments.

> To modify a predefined partition, first remove the partition from :file:`partitions.yml` then ensure that the build strategy for that sub image is "Build from source" and update the configured partition size.

Reviewer: ```suggestion
To modify a predefined partition, first remove the partition from :file:`partitions.yml` then ensure that the build strategy for that sub-image is "Build from source" and update the configured partition size.
```

> Note that only partitions adjacent with the ``app`` partition can be modified.

Reviewer 2: Note that only partitions adjacent to the ``app`` partition or other modified/deleted partitions can itself be modified.

Author: I will only state "adjacent to the app" here, because one will read the "remove" section to figure out how to make it adjacent. Pointing to removed partitions in this section will only be confusing IMO.

Reviewer 2: Ok, with the new comment referring to the remove section, I would say the sentence isn't needed.

Adding a partition
~~~~~~~~~~~~~~~~~~

> New partitions are added when sub-images are included in the build who's partitions are not listed in the predefined configuration.

Reviewer 2: How about

"New partitions are added when they are included in the build, but not listed in the predefined configuration."

> When using predefined configuration, the flash area available for new partitions is that of the ``app`` partition in addition

Reviewer 2: Is the app always re-solved?

Author: Yes, when predefined argument for PM is set.

to any unused area (for instance from deleted predefined partitions).

> Only partitions which are **not** included in :file:`partitions.yml` can be used as a reference for new partition definitions.

Reviewer 2: Not quite sure what this sentence is saying.

"used as a reference": "reference" is not a term used anywhere else in the documentation for predefined configuration AFAICT.

Author: Updated to 

`Only partitions which are **not** included in :file:`partitions.yml` can be used in the ``placement`` property for new partition definitions.`

Reviewer 2: Only partitions which are **not** included in :file:`partitions.yml` can be used in the ``placement: before`` or ``placement: after`` property for new partition definitions.

?

Author: (Y)

If desired, the file can be modified as described below.

.. _ug_pm_predefined_remove

Removing a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
A predefined partition is removed by deleting its entry in :file:`partitions.yml`.
Only partitions adjacent to the ``app`` partition or other removed partitions can itself be removed.

.. _ug_pm_predefined_modify

Modifying a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To modify a predefined partition, remove the partition from :file:`partitions.yml` as described above.
This will add its reserved area to the dynamic partition.
Now the partition will be configured by Partition Manager.
If the partition being modified is an image partition, ensure that the build strategy for corresponding sub-image is "Build from source" and update the configured partition size.
Only partitions adjacent with the ``app`` partition can be modified.

.. _ug_pm_predefined_add_dynamic

Adding a partition to dynamic partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> New partitions are added to the dynamic partition when they are included in the build (i.e. their :file:`pm.yml` is given to the partition manager script), but not listed in the predefined configuration.

Reviewer 2: ```suggestion
New partitions are added to the dynamic partition when they are included in the build (i.e. listed in a :file:`pm.yml` file.), but not listed in the predefined configuration.
```

> Only partitions which are **not** included in :file:`partitions.yml` can be used in the ``placement: before`` or ``placement: after`` property for new partition definitions.

Reviewer 2: ```suggestion
.. note::
   Partitions that are included in :file:`partitions.yml` will be ignored if found in the ``placement: before`` or ``placement: after`` property for new partition definitions.
```

.. _ug_pm_predefined_add

Adding a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To add a new predefined partition, add an entry for it in the file :file:`partitions.yml`.

> This entry must define the following properties:

Reviewer 2: For clarity, maybe list the properties in addition to giving the example.

Modifying a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To modify a predefined partition, remove the partition from :file:`partitions.yml` as described above.
This will add its reserved area to the dynamic partition.
Now the partition will be configured by Partition Manager.
If the partition being modified is an image partition, ensure that the build strategy for corresponding sub-image is "Build from source" and update the configured partition size.
Only partitions adjacent with the ``app`` partition can be modified.

.. _ug_pm_predefined_add_dynamic

Adding a partition to dynamic partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
New partitions are added to the dynamic partition when they are included in the build (i.e. listed in a :file:`pm.yml` file.), but not listed in the predefined configuration.
.. note::
   Partitions that are included in :file:`partitions.yml` will be ignored if found in the ``placement: before`` or ``placement: after`` property for new partition definitions.

.. _ug_pm_predefined_add

Adding a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To add a new predefined partition, add an entry for it in the file :file:`partitions.yml`.

> This entry must define the properties `address`, `size`, and if applicable - `span`.

Reviewer 2: Double backticks?

> .. code-block:: yaml

Reviewer 2: Add a small caption.

      --slot-size $<TARGET_PROPERTY:partition_manager,MCUBOOT_SLOT_SIZE>

.. _ug_pm_predefined:

Predefined configuration
========================
Predefined configuration is a feature intended for deployed products which consists of multiple images where
only a subset of the included images will be upgradeable through firmware update mechanisms.
One example of this is a device which includes a non-upgradable first stage bootloader and an upgradeable application.

> In these use cases, the images to be upgraded must be linked to the same address as that which is deployed.

Reviewer 3: If you only have 2 images, as this use-case appears to describe, and one is static, isn't the whole thing static?

Maybe we should use a use-case where there are two dynamic partitions?

Author: There can only exist one dynamic partition, I will try to explain more clearly.

Author: Changed name from dynamic partition to dynamic area to better explain that this is not some actual partition being handled like any other. There exists only one dynamic area, as described further up in this file with the updates.

> This to avoid accidental overrides, and incorrect configurations.

Reviewer 3: s/This to/This is to/

Also, remove the comma, a bit hypocrytical, I know.

> Predefined configuration allows the user to configure the **dynamic partition**.

Reviewer 3: Since we are using the word dynamic it seems unnatural to not use the word static as it's counterpart.

Author: Hm, using the word "partition" here might be confusing. It is only supposed to describe the area within which changes may take place (resizing etc). Maybe "dynamic area" is better? 

I struggle to find a good term to describe this so any proposals are welcome.

Author: Updated to "dynamic area" in which case I don't see the need to explicitly describe the static counterpart.

> Predefined configuration allows the user to configure the **dynamic partition**.

Reviewer 3: s/Predefined configuration allows/Predefined configuration still allows/

The dynamic partition consists of all memory adjacent to the "app" partition which is not occupied by a predefined partition.
When Partition Manager is executed, it only operates on the dynamic partition, assuming that all other memory is reserved.
Within the dynamic partition, it is allowed to define new partitions.
The dynamic partition is re-sized by adding or removing partitions to the predefined configuration.

.. _ug_pm_predefined_providing:

Providing predefined configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The predefined configuration is provided through a YAML formatted configuration file.
This file is similar to the regular :file:`pm.yml` configuration files, except

> that it also defines the start address and size of all partitions.

Reviewer 3: Is it permitted for the regular pm.yml file to set the start and size? If so we could say that "it must define the start address and size of all partitions."

Author: The regular pm.yml states the size, but not the start address. Will change it to 

`... also defines the start address of all partitions.`

> The given configuration for a build can be found in :file:`${BUILD_DIR}/zephyr/include/generated/partitions.yml`.

Reviewer 3: I believe it would be best to use a different filename for the current configuration and the predefined configuration.

Reviewer 3: prj.partitions.yml ?

Author: Good idea, will fix

When the build system sees a file named :file:`prj.partitions.yml` in an applications source directory, it is automatically provided to the partition manager script as the predefined configuration.
If desired, the file can be modified as described below.

.. _ug_pm_predefined_remove:

Removing a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
A predefined partition is removed by deleting its entry in :file:`prj.partitions.yml`.
Only partitions adjacent to the ``app`` partition or other removed partitions can itself be removed.

.. _ug_pm_predefined_modify:

Modifying a predefined partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To modify a predefined partition, remove the partition from :file:`prj.partitions.yml` as described above.
This will add its reserved area to the dynamic area.
Now the partition will be configured by Partition Manager.
If the partition being modified is an image partition, ensure that the build strategy for corresponding sub-image is "Build from source" and update the configured partition size.
Only partitions adjacent with the ``app`` partition can be modified.

.. _ug_pm_predefined_add_dynamic:

> Adding a partition to dynamic area

Reviewer 3: s/to dynamic area/to the dynamic area/

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
New partitions are added to the dynamic partition when they are included in the build (i.e. listed in a :file:`pm.yml` file.), but not listed in the predefined configuration.

.. note::

>    Partitions that are included in :file:`prj.partitions.yml` will be ignored if found in the ``placement: before`` or ``placement: after`` property for new partition definitions.

Reviewer 3: I'm worried about silently ignoring user-input.

Is there a use-case for having partitions in the prj.partitions.yml that are ignored?

Could we error-out instead?

Author: This is the same behavior we have in the normal pm.yml, which is really the same behavior. Note that this is not exactly "user input" as it is typically the author of the sub-image which describe its pm.yml.

I would argue that we have that use case, because one want to use the generic pm.yml provided by the author of the sub-image. This will have several partitions listed to be compatible with different configurations (imagine mcuboot tackling both with and without SPM partition).

Reviewer 3: I thought one would typically copy partitions.yml, not pm.yml, into prj.partitions.yml.

Then the user will anyway have a custom version.

Reviewer 3: But nevertheless, I can see that one would want Kconfig+pm.yml to have precedence over prj.partitions.yml.

Author: A user can combine both prj.partitions.yml and pm.yml.

.. _ug_pm_static_remove:

Removing a static partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~
A static partition is removed by deleting its entry in :file:`pm_static.yml`.
Only partitions adjacent to the ``app`` partition or other removed partitions can itself be removed.

.. _ug_pm_static_add_dynamic:

Adding a dynamic partition
~~~~~~~~~~~~~~~~~~~~~~~~~~
New dynamic partitions are added automatically by the build process if they are listed in a :file:`pm.yml` file, unless also listed in the static configuration.

.. note::
   When resolving the relative placement of dynamic partitions any placement
   properties referencing static partitions will be ignored.

.. _ug_pm_static_add:

Adding a static partition
~~~~~~~~~~~~~~~~~~~~~~~~~
To add a new static partition, add an entry for it in the file :file:`pm_static.yml`.
This entry must define the properties ``address``, ``size``, and if applicable - ``span``.

.. code-block:: yaml
   :caption: Example of static configuration of partition with span.

   partition_name:
      address: 0xab00
      size: 0x1000
      span: [example]  # Only if this partition had the `span` property set originally.

.. note::

>   Sub-images with the built strategy `*_BUILD_STRATEGY_SKIP_BUILD` or

Reviewer 3: ```Suggestion
  Sub-images with the build strategy `*_BUILD_STRATEGY_SKIP_BUILD` or
```

  `*_BUILD_STRATEGY_USE_HEX_FILE` must define a static partition to ensure correct

>   placement of the dynamic partitions.

Reviewer 3: What about images that define static partitions, do they need to enable _SKIP_BUILD or _USE_HEX_FILE ?

E.g. is it "a if b", or "a iff b"

Author: No, you are free to have any build strategy for static partitions, only their placement and size is set.

that it also defines the start address for all partitions.
The given configuration for a build can be found in :file:`${BUILD_DIR}/partitions.yml`.
To apply the current configuration as a static configuration, copy this file to `${APPLICATION_SOURCE_DIR}/pm_static.yml``.

.. note::
  Configuration of the `app` partition inside the static configuration is ignored.

When the build system sees a file named :file:`pm_static.yml` in an applications source directory, it is automatically provided to the partition manager script as the static configuration.
If desired, the file can be modified as described below.

.. _ug_pm_static_remove:

Removing a static partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~
A static partition is removed by deleting its entry in :file:`pm_static.yml`.
Only partitions adjacent to the ``app`` partition or other removed partitions can itself be removed.

.. _ug_pm_static_modify:

Modifying a static partition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
To modify a static partition, remove the partition from :file:`pm_static.yml` as described above.
This will add its reserved area to the dynamic area.
Now the partition will be configured by Partition Manager.

> If the partition being modified is an image partition, ensure that the build strategy for corresponding sub-image is "Build from source" and update the configured partition size.

Reviewer 3: s/for corresponding/for the corresponding/
